Webapp/tests/services/test_decorators.py

```diff
@@ -0,0 +1,122 @@
+from sources.decorators import (
+    principal_investigator_required,
+    principal_investigator_or_senior_researcher_required,
+    workbook_member_required,
+    workgroup_member_required,
+    _is_demo,
+    _get_from_request
+)
+
+
+@pytest.mark.parametrize("input_value, search_str, request_args, request_form, expected", [
+    (None, "workgroup", {"workgroup": "arg_value"}, {}, "arg_value"),
+    (None, "workgroup", {}, {"workgroup": "form_value"}, "form_value"),
+    ("default_value", "workgroup", {}, {}, "default_value"),
+])
+def test_get_from_request(input_value, search_str, request_args, request_form, expected, app):
+    with app.test_request_context(query_string=request_args, data=request_form):
+        assert _get_from_request(input_value, search_str) == expected
+
+
+@pytest.mark.parametrize(
+    "request_args, request_form, expected", [
+        ({"demo": "demo"}, {}, True),
+        ({"tutorial": "yes"}, {}, True),
+        ({}, {"demo": "demo"}, True),
+        ({}, {"tutorial": "yes"}, True),
+        ({}, {}, False),
+    ]
+)
+def test_is_demo(request_args, request_form, expected, app):
+    with app.test_request_context(query_string=request_args, data=request_form):
+        assert _is_demo() == expected
+
+
+def mock_function(workgroup, workbook):
+    return jsonify("Access Granted")
+
+
+def setup_user_type_mocker(mocker: MockFixture, user_type: str):
+    mock_get_user_type = mocker.patch("sources.services.workgroup.get_user_type")
+    mock_get_user_type.return_value = user_type
+
+
+def setup_get_workgroups_mocker(mocker: MockFixture, workgroups: List[str]):
+    mock_get_workgroups = mocker.patch("sources.auxiliary.get_workgroups")
+    mock_get_workgroups.return_value = workgroups
+
+
+def check_response_success(app: Flask, decorated_function, workgroup="Test-Workgroup", workbook=None):
+    with app.test_request_context():
+        response = decorated_function(workgroup=workgroup, workbook=workbook)
+        assert response.status_code == 200
+        assert response.json == "Access Granted"
+
+
+def check_response_failed(app: Flask, decorated_function: mock_function, workgroup="Test-Workgroup", workbook=None):
+    with app.test_request_context():
+        response = decorated_function(workgroup=workgroup, workbook=workbook)
+        assert response.status_code == 302
+        assert response.location == url_for("main.index")
+
+
+def test_principal_investigator_required(client: FlaskClient, mocker: MockFixture, app: Flask):
+    decorated_function = principal_investigator_required(mock_function)
+
+    # test user is PI
+    setup_user_type_mocker(mocker, user_type="principal_investigator")
+    check_response_success(app, decorated_function)
+
+    # test user is not PI
+    setup_user_type_mocker(mocker, user_type="not_PI")
+    check_response_failed(app, decorated_function)
+
+
+def test_principal_investigator_or_senior_researcher_required(client: FlaskClient, mocker: MockFixture, app: Flask):
```

Comment:
The tests should probably have docstrings to explain what is being tested

Reply:
Done